File: Dev/C++/Mascaret/src/BEHAVE/Agent.cpp
#include "BEHAVE/Agent.h"
#include "BEHAVE/AgentClass.h"
#include "BEHAVE/AgentBehavior.h"
#include "BEHAVE/AgentBehaviorExecution.h"
#include "BEHAVE/SimpleBehaviorExecution.h"
#include "BEHAVE/SimpleCommunicationBehavior.h"
#include "BEHAVE/ProceduralBehavior.h"
#include "BEHAVE/ProcedureExecution.h"
#include "BEHAVE/OneShotBehaviorExecution.h"
#include "BEHAVE/AgentPlateform.h"
#include "BEHAVE/ACLMessage.h"
#include "VEHA/Behavior/Common/behaviorScheduler.h"
#include "VEHA/Behavior/Common/Signal.h"
#include "VEHA/Behavior/BasicActions/Action.h"
#include "VEHA/Behavior/Activity/Activity.h"
#include "VEHA/Behavior/Activity/ActivityNode.h"
#include "VEHA/Behavior/Activity/ActionNode.h"
#include "VEHA/Behavior/BasicActions/SendSignalAction.h"
#include "VEHA/Kernel/Operation.h"

namespace BEHAVE
{
using VEHA::BehaviorScheduler;
Agent::Agent(shared_ptr<AgentPlateform> ap, const string & name, shared_ptr<AgentClass> agentclass, string KBName) : InstanceSpecification(name, agentclass), _aid(shared_ptr<AID>(new AID(name,ap->getName(), ap->getPort()))), _plateform(ap),_mailbox(shared_ptr<Mailbox>(new Mailbox())), _knowledgeBase(shared_ptr<KnowledgeBase>(new KnowledgeBase(KBName)))
{
}

Agent::~Agent()
{
}

void Agent::addBehavior(const string& behaviorName,double interval, bool start)
{
	if(getBehaviorByName(behaviorName))
		return;

	shared_ptr<AgentBehavior> behavior(new AgentBehavior(behaviorName));
	Parameters param;
	// TODO: le comportement est créer 2 fois. Il doit y avoir mieux a faire...
	shared_ptr<BehaviorExecution> be=behavior->createBehaviorExecution(shared_dynamic_cast<InstanceSpecification>(shared_from_this()),param);
	if(!shared_dynamic_cast<OneShotBehaviorExecution>(be) && start)
	{
		be = BehaviorScheduler::getInstance()->executeBehavior(shared_dynamic_cast<Behavior>(behavior),shared_dynamic_cast<InstanceSpecification>(shared_from_this()),param);
		if(interval > 0)
		{
			shared_ptr<SimpleBehaviorExecution> sbe=shared_dynamic_cast<SimpleBehaviorExecution>(be);
			if(sbe)
				sbe->setInterval(interval);
		}
		_behaviorsExecution.push_back(shared_dynamic_cast<AgentBehaviorExecution>(be));
	}
	_behaviors.push_back(behavior);
}

shared_ptr<AgentBehavior> Agent::getBehaviorByName(const string& behaviorName)
{
	shared_ptr<AgentBehavior> behavior;

	for (size_t i = 0; i < _behaviors.size(); i++)
	{
		if (_behaviors[i]->getName() == behaviorName)
			behavior = _behaviors[i];
	}

	return behavior;
}
shared_ptr<AgentBehaviorExecution> Agent::executeOneShotBehavior(const string& behaviorName)
{
	shared_ptr<AgentBehavior> behavior=getBehaviorByName(behaviorName);
	Parameters param;
	// TODO: le comportement est créer 2 fois. Il doit y avoir mieux a faire...
	shared_ptr<BehaviorExecution> be=behavior->createBehaviorExecution(shared_dynamic_cast<InstanceSpecification>(shared_from_this()),param);
	if(shared_dynamic_cast<OneShotBehaviorExecution>(be))
	{
		be = BehaviorScheduler::getInstance()->executeBehavior(shared_dynamic_cast<Behavior>(behavior),shared_dynamic_cast<InstanceSpecification>(shared_from_this()),param);
		return shared_dynamic_cast<AgentBehaviorExecution>(be);
	}
	return shared_ptr<AgentBehaviorExecution>();
}
shared_ptr<AgentBehaviorExecution> Agent::getBehaviorExecutionByName(const string& behaviorName)
{
	shared_ptr<AgentBehaviorExecution> behavior;

	for (size_t i = 0; i < _behaviorsExecution.size(); i++)
	{
		if (_behaviorsExecution[i]->getSpecification()->getName() == behaviorName)
			behavior = _behaviorsExecution[i];
	}

	return behavior;
}


void Agent::postMessage(shared_ptr<ACLMessage> message)
{
	_mailbox->postMessage(message);
}

shared_ptr<ACLMessage> Agent::receive()
{
	return _mailbox->check();
}

void Agent::send(shared_ptr<ACLMessage> message)
{
	message->setSender(getAID());
	getAgentPlateform()->sendMessage(message);
	_mailbox->send(message);
}

void Agent::addExecution(shared_ptr<BehaviorExecution> be)
{
	InstanceSpecification::addExecution(be);

	shared_ptr<Behavior> behaviorDone = be->getSpecification();

	shared_ptr<Class> classe = getClassifier();
	if (!classe) return;

	map<string, shared_ptr<Operation> > operations = classe->getOperations();

	bool found = false;
	shared_ptr<Operation> operationDone;
	map<string,  shared_ptr<Operation> >::iterator it = operations.begin();
	while( it != operations.end() && !found)
	{
		shared_ptr<Behavior> operationBehavior = it->second->getMethod();
		if (operationBehavior)
		{
			shared_ptr<Activity> act = shared_dynamic_cast<Activity>(operationBehavior);
			if (act)
			{
				shared_ptr<ActivityNode> init = act->getInitialNode();

				vector< shared_ptr<ActionNode> > ans = init->getOutgoingActionNode();
				if (ans.size() > 0)
				{
					shared_ptr<ActionNode> actionNode = ans[0];
					shared_ptr<VEHA::Action> action = actionNode->getAction();
					shared_ptr<SendSignalAction> sendSigAction = shared_dynamic_cast<SendSignalAction>(action);	
					if (sendSigAction)
					{
						shared_ptr<SendSignalAction> sendSigDone = shared_dynamic_cast<SendSignalAction>(behaviorDone);
						if (sendSigDone)
						{
							if (sendSigDone->getSignalClass()->getName() == sendSigAction->getSignalClass()->getName())
							{
								string targetClassNameDone = sendSigDone->getTarget()->getClassifier()->getName();
								vector< shared_ptr<ObjectNode> > objs = actionNode->getOutgoingObjectNode();
								if (objs.size() > 0)
								{
									if (objs[0]->getResourceType()->getName() == targetClassNameDone)
									{
										found = true;
										operationDone = it->second;
									} 
								}
							}
						}
					}
				}
			}
		}
		it ++;
	}

	if (found)
	{
		shared_ptr<AgentBehaviorExecution> abe = getBehaviorExecutionByName("SimpleCommunicationBehavior");
                shared_ptr<SimpleCommunicationBehavior> cbe  = shared_dynamic_cast<SimpleCommunicationBehavior> (abe);
		vector<shared_ptr<ActionNode> > possibleTranslation = cbe->_translateOperationToActions(operationDone);		

		shared_ptr<AgentBehaviorExecution> pbe = getBehaviorExecutionByName("ProceduralBehavior"); 
                shared_ptr<ProceduralBehavior> pb  = shared_dynamic_cast<ProceduralBehavior> (pbe);
		if (possibleTranslation.size() >0 && pb)
		{
			shared_ptr<ActionNode> an = possibleTranslation[0];
		        cerr << " # Action : " << an->getName() << endl;
			an->_currentExecution = be;
			vector<shared_ptr<ProcedureExecution> > procs = pb->accessRunningProcedures();
			shared_ptr<ProcedureExecution> procInfo = procs[0];
			pb->_sendActionRealisationMessage(an, procInfo); // TODO
                        procInfo->informActionRunning(getAID(),an->getName());

			procInfo->informActionDone(getAID(),an->getName()); 
                        pb->_sendActionDoneMessage(an, procInfo);
		}

	}
	
}

void Agent::wakeup()
{
	for (size_t i = 0; i < _behaviorsExecution.size(); i++)
	{
		if (_behaviorsExecution[i] && _behaviorsExecution[i]->isSuspended())
		{
			_behaviorsExecution[i]->restart();
		}
	}
}

}

File: Dev/CS/Mascaret/Mascaret/BEHAVE/Agent.cs
using System;
using System.Collections.Generic;

namespace Mascaret
{
    public class Agent : InstanceSpecification
    {
        public string emotion;

        private List<AgentBehavior> behaviors = new List<AgentBehavior>();
        public List<AgentBehavior> Behaviors
        {
            get { return behaviors; }
            set { behaviors = value; }
        }


        private List<AgentBehaviorExecution> behaviorsExecution = new List<AgentBehaviorExecution>();
        public List<AgentBehaviorExecution> BehaviorsExecution
        {
            get { return behaviorsExecution; }
            set { behaviorsExecution = value; }
        }


        private AID aid;
        public AID Aid
        {
            get { return aid; }
            set { aid = value; }
        }


        private AgentPlateform plateform;
        public AgentPlateform Plateform
        {
            get { return plateform; }
            set { plateform = value; }
        }


        private Mailbox mailbox;
        public Mailbox Mailbox
        {
            get { return mailbox; }
            set { mailbox = value; }
        }


        private KnowledgeBase knowledgeBase;
        public KnowledgeBase KnowledgeBase
        {
            get { return knowledgeBase; }
            set { knowledgeBase = value; }
        }


        private bool controlledByHuman;
        public bool ControlledByHuman
        {
            get { return controlledByHuman; }
            set { controlledByHuman = value; }
        }

        public BehaviorPlanner behaviorPlanner;

        //default parameters KBName = "default"
        public Agent(AgentPlateform ap, string name, AgentClass agentclass, string KBName)
            : base(name, agentclass)
        {

            this.aid = new AID(name, ap.name, ap.PortNumber);
            this.plateform = ap;
            mailbox = new Mailbox();
            knowledgeBase = new KnowledgeBase(KBName);
            //behaviorPlanner = new BehaviorPlanner(this);
            //behaviorRealizer = new BehaviorRealizer();
            emotion = "joyStrong";
        }

        //default parameters interval = -1, start = true
        public void addBehavior(string behaviorName, double interval, bool start)
        {
            if (getBehaviorByBame(behaviorName) != null)
                return;

            AgentBehavior behavior = new AgentBehavior(behaviorName);
            Dictionary<string, ValueSpecification> param = new Dictionary<string, ValueSpecification>();


            BehaviorExecution be = behavior.createBehaviorExecution((InstanceSpecification)this, param, false);
            //if(be as OneShotBehaviorExecution !=null && start)
            {
                be = BehaviorScheduler.Instance.executeBehavior((Behavior)behavior, (InstanceSpecification)this, param, false);
                if (interval > 0.00)
                {
                    SimpleBehaviorExecution sbe = (SimpleBehaviorExecution)be;
                    if (sbe != null)
                        sbe.Interval = interval;
                }
                behaviorsExecution.Add((AgentBehaviorExecution)be);
            }
            behaviors.Add(behavior);

        }

        public AgentBehavior getBehaviorByBame(string behaviorName)
        {
            foreach (AgentBehavior currentAgentBehavior in behaviors)
            {
                if (currentAgentBehavior.name.ToLower() == behaviorName.ToLower())
                    return currentAgentBehavior;
            }
            return null;
        }

        public AgentBehaviorExecution executeOneShotBehavior(string behaviorName)
        {
            AgentBehavior behavior = getBehaviorByBame(behaviorName);
            Dictionary<string, ValueSpecification> param = new Dictionary<string, ValueSpecification>();

            BehaviorExecution be = behavior.createBehaviorExecution((InstanceSpecification)this, param, false);
            if (be as OneShotBehaviorExecution != null)
            {
                be = BehaviorScheduler.Instance.executeBehavior((Behavior)behavior, (InstanceSpecification)this, param, false);
                return (AgentBehaviorExecution)be;
            }
            return null;
        }

        public AgentBehaviorExecution getBehaviorExecutingByName(string behaviorName)
        {
            System.Console.WriteLine(this.name + " : " + behaviorsExecution.Count);
            foreach (AgentBehaviorExecution currentBe in behaviorsExecution)
            {
                if (currentBe != null)
                    if (currentBe.Specification.name == behaviorName)
                        return currentBe;
            }
            return null;
        }

        public void postMessage(ACLMessage message)
        {
            this.mailbox.postMessage(message);
        }


        public ACLMessage receive()
        {
            return this.Mailbox.check();
        }


        public void send(ACLMessage message)
        {
            message.Sender = this.aid;
            this.plateform.sendMessage(message);
            this.mailbox.send(message);
        }

        public List<ACLMessage> getMessageQueue()
        {
            return mailbox.MessagesQueue;
        }

        public List<ACLMessage> getChecked()
        {
            return mailbox.MessagesChecked;
        }

        public List<ACLMessage> getSent()
        {
            return mailbox.MessagesSent;
        }

        //bilal 19-10-15
        public virtual void addIntention(string fml)
        {
            behaviorPlanner.parseIntention(fml);
        }
        //bilal 19-10-15

        public void wakeup()
        {
        }

        public virtual void start()
        {
            Dictionary<string, ValueSpecification> param = new Dictionary<string, ValueSpecification>();
            foreach (Behavior behavior in this.Behaviors)
            {
                BehaviorExecution be = BehaviorScheduler.Instance.executeBehavior((Behavior)behavior, (InstanceSpecification)this, param, false);
                behaviorsExecution.Add((AgentBehaviorExecution)be);
            }

        }



    }
}

